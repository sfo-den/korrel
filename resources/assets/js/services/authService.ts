import { merge } from 'lodash'
import { http, localStorageService } from '@/services'
import { userStore } from '@/stores'

export interface UpdateCurrentProfileData {
  current_password: string | null
  name: string
  email: string
  avatar?: string
  new_password?: string
}

interface CompositeToken {
  'audio-token': string
  'token': string
}

const API_TOKEN_STORAGE_KEY = 'api-token'
const AUDIO_TOKEN_STORAGE_KEY = 'audio-token'

export const authService = {
  async login (email: string, password: string) {
    const token = await http.post<CompositeToken>('me', { email, password })

    this.setAudioToken(token['audio-token'])
    this.setApiToken(token.token)
  },

  async logout () {
    await http.delete('me')
    this.destroy()
  },

  getProfile: async () => await http.get<User>('me'),

  updateProfile: async (data: UpdateCurrentProfileData) => {
    merge(userStore.current, (await http.put<User>('me', data)))
  },

  getApiToken: () => localStorageService.get(API_TOKEN_STORAGE_KEY),

  hasApiToken () {
    return Boolean(this.getApiToken())
  },

  setApiToken: (token: string) => localStorageService.set(API_TOKEN_STORAGE_KEY, token),

  destroy: () => {
    localStorageService.remove(API_TOKEN_STORAGE_KEY)
    localStorageService.remove(AUDIO_TOKEN_STORAGE_KEY)
  },

  setAudioToken: (token: string) => localStorageService.set(AUDIO_TOKEN_STORAGE_KEY, token),

  getAudioToken: () => {
    // for backward compatibility, we first try to get the audio token, and fall back to the (full-privileged) API token
    return localStorageService.get(AUDIO_TOKEN_STORAGE_KEY) || localStorageService.get(API_TOKEN_STORAGE_KEY)
  }
}
