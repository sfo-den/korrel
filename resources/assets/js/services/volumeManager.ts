import { ref } from 'vue'
import { preferenceStore } from '@/stores'

export class VolumeManager {
  private input!: HTMLInputElement
  public volume = ref(0)

  public init (input: HTMLInputElement) {
    this.input = input
    this.set(preferenceStore.volume)
  }

  public get () {
    return this.volume.value
  }

  public set (volume: number) {
    this.volume.value = volume
    this.input.value = String(volume)
  }

  public mute () {
    this.set(0)
  }

  public unmute () {
    this.set(preferenceStore.volume)
  }
}

export const volumeManager = new VolumeManager()
