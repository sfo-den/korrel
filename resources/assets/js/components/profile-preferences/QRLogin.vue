<template>
  <article class="text-secondary">
    Instead of using a password, you can scan the QR code below to log in to
    <a href="https://koel.dev/#mobile" target="_blank" class="text-highlight">Koel Player</a>
    on your mobile device.<br>
    The QR code will refresh every 10 minutes.
    <img :src="qrCodeUrl" alt="QR Code" width="192" height="192">
  </article>
</template>

<script setup lang="ts">
import { onMounted, ref, watch } from 'vue'
import { useQRCode } from '@vueuse/integrations/useQRCode'
import { authService } from '@/services'
import { base64Encode } from '@/utils'

const qrCodeData = ref('')
const oneTimeToken = ref('')

watch(oneTimeToken, () => {
  qrCodeData.value = base64Encode(JSON.stringify({
    token: oneTimeToken.value,
    url: window.BASE_URL
  }))
})

const qrCodeUrl = useQRCode(qrCodeData, {
  width: window.devicePixelRatio === 1 ? 196 : 384,
  height: window.devicePixelRatio === 1 ? 196 : 384
})

const resetOneTimeToken = async () => (oneTimeToken.value = await authService.getOneTimeToken())

onMounted(() => {
  window.setInterval(resetOneTimeToken, 60 * 10 * 1000)
  resetOneTimeToken()
})
</script>

<style scoped lang="postcss">
img {
  margin-top: 1.5rem;
  display: block;
  border-radius: 8px;
}
</style>
