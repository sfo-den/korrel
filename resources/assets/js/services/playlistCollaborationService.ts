import { http } from '@/services'

export const playlistCollaborationService = {
  async createInviteLink (playlist: Playlist) {
    if (playlist.is_smart) {
      throw Error('Smart playlists are not collaborative.')
    }

    const token = (await http.post<{ token: string }>(`playlists/${playlist.id}/collaborators/invite`)).token
    return `${window.location.origin}/#/playlist/collaborate/${token}`
  },

  async acceptInvite (token: string) {
    return http.post<Playlist>(`playlists/collaborators/accept`, { token })
  }
}
