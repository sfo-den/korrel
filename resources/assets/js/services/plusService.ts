import { http } from '@/services'

export const plusService = {
  activateLicense: async (key: string) => {
    return await http.post('licenses/activate', { key })
  }
}
