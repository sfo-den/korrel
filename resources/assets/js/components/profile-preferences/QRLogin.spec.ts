import { expect, it, vi } from 'vitest'
import UnitTestCase from '@/__tests__/UnitTestCase'
import { authService } from '@/services'
import QRLogin from '@/components/profile-preferences/QRLogin.vue'

new class extends UnitTestCase {
  protected test () {
    it('renders', async () => {
      vi.mock('@vueuse/integrations/useQRCode', () => ({
        useQRCode: () => 'data:image/png;base64,my-qr-code'
      }))

      const getTokenMock = this.mock(authService, 'getOneTimeToken').mockResolvedValue('my-token')
      const { html } = this.render(QRLogin)

      expect(getTokenMock).toHaveBeenCalled()
      expect(html()).toMatchSnapshot()
    })
  }
}
