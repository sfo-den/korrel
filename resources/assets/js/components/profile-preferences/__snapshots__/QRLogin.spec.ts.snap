// Vitest Snapshot v1

exports[`renders 1`] = `<article class="text-k-text-secondary"> Instead of using a password, you can scan the QR code below to log in to <a class="text-k-highlight" href="https://koel.dev/#mobile" target="_blank">Koel Player</a> on your mobile device.<br> The QR code will refresh every 10 minutes. <img src="data:image/png;base64,my-qr-code" alt="QR Code" class="mt-4 rounded-4" height="192" width="192"></article>`;
